File: server/domain/dto/scrap.py
from typing import List, Optional

from pydantic import BaseModel

from server.domain.dto.image import ImageResponse

from ..entity import Scrap, Tags
from ..PageType import PageType


class URLInput(BaseModel):
    url: str


class ScrapModifier(BaseModel):
    content: str
    author_name: str
    author_tag: str
    pin: Optional[bool] = None
    comment: Optional[str] = None
    tags: Optional[List[str]] = None


class ScrapCreate(ScrapModifier):
    source_key: str
    source: str
    image_names: List[str]


class ScrapUpdate(ScrapModifier):
    delete_images: list[int]
    new_images: list[int]


class ScrapResponse(BaseModel):
    id: int
    source: str
    source_id: str
    url: str
    content: Optional[str]
    author_name: str
    author_tag: str
    comment: Optional[str]
    images: List[ImageResponse]
    pin: bool
    tags: list[str] = []

    @staticmethod
    def fromScrap(scrap: Scrap, tags: list[Tags] = None):
        if tags is None:
            tags = []

        return ScrapResponse(
            id=scrap.id,
            source_id=scrap.source_id,
            content=scrap.content,
            comment=scrap.comment,
            url=PageType[scrap.source].url(scrap.source_id),
            author_name=scrap.author_name,
            author_tag=scrap.author_tag,
            images=[ImageResponse(id=image.id, file_name=image.file_name) for image in scrap.images],
            source=scrap.source,
            pin=scrap.pin,
            tags=[tag.name for tag in tags],
        )

    def setTag(self, tags: list[Tags]):
        self.tags.extend([tag.name for tag in tags])

File: internal/model/dto/scrap.go
package dto

import (
    "Scrapper/internal/model/entity"
    "Scrapper/internal/model/entity/pageType"
)

type URLInput struct {
    Url string
}

type ScrapModifier struct {
    Content    *string
    AuthorName string
    AuthorTag  string
    Pin        *bool
    Comment    *string
    Tags       *[]string
}

type ScrapCreate struct {
    ScrapModifier
    SourceKey  string
    Source     pageType.PageType
    ImageNames []string
}

type ScrapUpdate struct {
    ScrapModifier
    DeleteImages []int
    NewImages    []int
}

func Create2Update(create ScrapCreate) ScrapUpdate {
    return ScrapUpdate{
        ScrapModifier: ScrapModifier{
            Content:    create.Content,
            AuthorName: create.AuthorName,
            AuthorTag:  create.AuthorTag,
            Pin:        create.Pin,
            Comment:    create.Comment,
            Tags:       create.Tags,
        },
    }
}

type ScrapResponse struct {
    Id         uint
    Source     pageType.PageType
    SourceId   string
    Url        string
    Content    *string
    AuthorName string
    AuthorTag  string
    Comment    *string
    Images     []ImageResponse
    Pin        bool
    Tags       []string
}

func NewScrapResponse(scrap entity.Scrap, tags []entity.Tag) ScrapResponse {
    tagNames := make([]string, len(tags))
    for i, tag := range tags {
        tagNames[i] = tag.Name
    }

    images := make([]ImageResponse, len(scrap.Images))
    for i, img := range scrap.Images {
        images[i] = ImageResponse{img.ID, img.FileName}
    }

    return ScrapResponse{
        Id:         scrap.ID,
        SourceId:   scrap.SourceID,
        Url:        scrap.Source.Url(scrap.SourceID),
        Content:    scrap.Content,
        AuthorName: scrap.AuthorName,
        AuthorTag:  scrap.AuthorTag,
        Comment:    scrap.Comment,
        Images:     images,
        Source:     scrap.Source,
        Pin:        scrap.Pin,
        Tags:       tagNames,
    }
}
